File Transfer RFD (#11526)
gravitational/teleport: rfd/0058-desktop-file-transfer.md

The data flow when using the clipboard virtual channel extension (henceforth referred to as RDP Option 1) includes the one described in [Data Flow and Delayed Rendering](https://github.com/gravitational/teleport/blob/master/rfd/0049-desktop-clipboard.md#data-flow-and-delayed-rendering) in RFD 0049, but with extra steps appended. In the case that a file, directory, or multiple files/directories have been cut or copied onto the shared clipboard, the `Format Data Response PDU` doesn't respond with the file data itself, but instead sends a [`Packed File List`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/3570c2e4-cdd7-4460-8a7e-1a4595f5ebdc) which contains a list of filenames and associated metadata of the files on the clipboard. The files themselves aren't transferred until the shared clipboard owner receives a [`File Contents Request PDU`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/cbc851d3-4e68-45f4-9292-26872a9209f2), at which point it sends back the file data in the form of [`File Contents Response PDU`(`s`)](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/df87c178-ab02-471a-acde-bb921aa1af85).

> File transfer's are limited to files less than 4GB unless huge file support is enabled (`CB_HUGE_FILE_SUPPORT_ENABLED`).

Reviewer: ```suggestion
File transfers are limited to files less than 4GB unless huge file support is enabled (`CB_HUGE_FILE_SUPPORT_ENABLED`).
```

### File System Virtual Channel Extension (RDP Option 2)

> From a high level, the file system virtual channel extension kicks off by the RDP client announcing a directory it wishes to share. From then on out, all functionality on that directory is instigated by the RDP server, carried out by the RDP client, and then if the operation on the client was successful, it is communicated back to and carried out on the server.

Reviewer: ```suggestion
From a high level, the file system virtual channel extension kicks off by the RDP client announcing a directory it wishes to share. From then on out, all actions on that directory are initiated by the RDP server, carried out by the RDP client, and then if the operation on the client was successful, it is communicated back to and carried out on the server.
```

Reworded this a bit, but it's still not quite clear to me what you're describing here.

> For example, if a program on the server wants to read a shared file, it sends a [`Device Read Request`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/3192516d-36a6-47c5-987a-55c214aa0441), and the client reads the requested data and sends it back with a [`Device Read Response`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/d35d3f91-fc5b-492b-80be-47f483ad1dc9). If instead the server-side program wants to write to a shared file, it sends a [`Device Write Request`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/2e25f0aa-a4ce-4ff3-ad62-ab6098280a3a) specifying where and what data it wants to write. In the happy path, the client writes that data to disk and sends back a [`Device Write Response`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/58160a47-2379-4c4a-a99d-24a1a666c02a) with the number of bytes written. If the client-side write fails, it sends back the `Device Write Response` with an `IoStatus` field set to something besides `STATUS_SUCCESS` (such as `STATUS_UNSUCCESSFUL`), which alerts the server make the write operation fail on its end as well.

Author: I'm not 100% sure. I had the same question, and tested that it does get notified by experimenting with drive redirection in the official Microsoft Remote Desktop client. One possibility is that it polls, another is that it re-requests metadata (file/directory names and sizes) every time some program on the server side requests access to the shared drive.

For our purposes, the file system virtual channel extension option (RDP Option 2) can be broken down into two sub options, RDP Option 2a and RDP Option 2b. RDP Option 2a is that we share a directory on the machine running the RDP client (`windows_desktop_service`), and RDP Option 2b is sharing a directory that lives on the Teleport user's client itself (via the File System Access API, discussed in more detail below).

## Client Side Possibilities

Irrespective of which option we choose, we will need some means of initiating and completing file transfers on the client side (from the browser).

### Clipboard API

At first glance, it seemed like using the `Clipboard.read` and `Clipboard.write` API's might be a possibile mechanism for us to make use of. However while some files can be transferred to the client's clipboard with this API, browsers are [constrained](https://w3c.github.io/clipboard-apis/#reading-from-clipboard) to the `text/plain` `text/html` and `image/png` mime types. Ergo unless we want to limit our users to a very narrow set of file transfers, this option is not available to us.

> ### Browser Upload and Download (Client Option 1)

Reviewer: I'm having a hard time keeping track of which option is being discussed due to the deep level of nesting here.

Is "Client Option 1" a sub-option of RDP Option 2a?

Author: Yeah, my naming system is questionable, in hindsight the numbering might be more confusing than helpful. I tried to clarify by the structure of the headings, but its difficult to follow, particularly reading the plain Markdown. 

I've done some re-naming and here is a summary of the structure based on the headings:
- **RDP Possibilities and Data Flow (RDP Options 1, 2a, and 2b)**
  - Clipboard Virtual Channel Extension (RDP Option 1)
  - File System Virtual Channel Extension (RDP Option 2a and 2b)
- **Client Side Possibilities (Client Options 1 and 2)**
  - Browser Upload and Download (Client Option 1)
  - TDP Upload and Download via the File System Access API (Client Option 2)
- **Implementation Options (Full Implementation Options 1, 2, 3, and 4)**
  - RDP Option 1 + Client Option 1 (Full Implementation Option 1)
  - RDP Option 1 + Client Option 2 (Full Implementation Option 2)
  - RDP Option 2a + Client Option 1 (Full Implementation Option 3)
  - RDP Option 2b + Client Option 2 (Full Implementation Option 4)

The first two **bolded bullet points** contain our options on the backend and frontend respectively, the last one lays out the plausible combinations of first two that are worth considering.

---
authors: Isaiah Becker-Mayer (isaiah@goteleport.com)
state: draft
---

> 

Reviewer 2: ```suggestion
# Required Approvers
* Engineering: @zmb3 && (@probakowski || @LKozlowski)
* Product: (@klizhentas || @xinding33)
```
